libyul/AST.h

```diff
@@ -41,7 +43,41 @@ using TypedNameList = std::vector<TypedName>;
 
 /// Literal number or string (up to 32 bytes)
 enum class LiteralKind { Number, Boolean, String };
-struct Literal { langutil::DebugData::ConstPtr debugData; LiteralKind kind; YulString value; Type type; };
+/// Literal value that holds a u256 word of data, can be of LiteralKind type and - in case of arguments to
+/// builtins - exceed the u256 word (32 characters), in which case the value is set to `unlimited`
+class LiteralValue {
+public:
+	using Data = u256;
+	using RepresentationHint = std::shared_ptr<std::string>;
+
+	LiteralValue() = default;
+	LiteralValue(Data const& _data, std::optional<std::string> const& _hint = std::nullopt, bool _unlimited = false):
+		m_data(_data), m_hint(_hint ? std::make_shared<std::string>(*_hint) : nullptr), m_unlimited(_unlimited) {}
```

Comment:
Can we assert here that the hint matches the data if present?

`formatLiteral()` ignores `m_data` when `m_hint` is present so we must ensure that these two are always consistent.

Reply:
And I'd recommend this style:
```suggestion
	LiteralValue(Data const& _data, std::optional<std::string> const& _hint = std::nullopt, bool _unlimited = false):
		m_data(_data),
		m_hint(_hint ? std::make_shared<std::string>(*_hint) : nullptr),
		m_unlimited(_unlimited)
	{}
```

Reply:
Hmm asserting it in this place is going to be difficult without knowing the LiteralKind. That is what determines the validity in the end. I have added a check to `formatLiteral` in the utilities, though, so that the hint isn't blindly taken but it is compared to the stored u256 value.

Alternatively, we could think about storing the `LiteralKind` in the `LiteralValue` as well. Then an assert in the constructor is possible and failure would occur earlier.

Reply:
Yeah, it really feels like `LiteralKind` should be inside `LiteralValue`, because the value is ambiguous without it. If would let us get rid of a lot of redundant asserts that we now have to do at every step to ensure the values are valid.

But I think it would require changing a lot more code so perhaps better to leave at is for now and just finish the PR in the current state.